[paimon connector]select error: Failed to call the nextChunkOffHeap method of off-heap table scanner

<!-- (At least include the following, feel free to add if you have more content) -->

### Steps to reproduce the behavior (Required)
1. CREATE PAIMON TABLE IN HIVE AND INSERT DATA:
```
create table paimon_hdfs_orc_map_nest_lz4(
 col_string string,
 col_map_map_map_map map<int,map<string,map<date,map<double,string>>>>,
 col_map_array_struct_map map<string,array<struct<c_date:date,c_string:string,c_map:map<boolean,date>>>>,
 col_map_array_map_struct map<boolean,array<map<string,struct<c_boolean:boolean,c_date:date>>>>,
 col_map_struct_map_array map<date,struct<c_map:map<date,array<boolean>>,c_date:date>>,
 col_map_struct_array_map map<tinyint,struct<c_date:date,c_array:array<map<boolean,date>>,c_boolean:boolean>>,
 col_map_map_struct_array map<decimal(5,2),map<int,struct<c_array:array<date>,c_string:string,c_bigint:bigint>>>,
 col_map_map_array_struct map<timestamp,map<int,array<struct<c_date:date,c_string:string,c_boolean:boolean>>>>
)
STORED BY 'org.apache.paimon.hive.PaimonStorageHandler'
TBLPROPERTIES (
    'bucket' = '2',
    'file.format' = 'orc',
    'file.compression' = 'lz4'
);

insert into paimon_hdfs_orc_map_nest_lz4 select 9,
 map(cast(1 as int),map('hello',map(cast('1999-12-31' as date),map(cast(3.1415926 as double),'China-Beijing',cast(3.1415927 as double),'China-Shanghai')))),
 map('America-LA',array(named_struct("c_date",cast('2020-02-02' as date),"c_string",'China-Chongqing',"c_map",map(cast(1 as boolean),cast('1010-10-10' as date),cast(0 as boolean),cast('1990-03-25' as date))))),
 map(cast(1 as boolean),array(map('China',named_struct("c_boolean",cast(1 as boolean),"c_date",cast('1010-01-01' as date))),map('Japan',named_struct("c_boolean",cast(0 as boolean),"c_date",cast('2020-02-02' as date))),map('America',named_struct("c_boolean",cast(null as boolean),"c_date",cast('3030-03-03' as date))))),
 map(cast('2020-02-02' as date),named_struct("c_map",map(cast('1010-01-01' as date),array(cast(1 as boolean),cast(0 as boolean),cast(0 as boolean),cast(1 as boolean),cast(null as boolean),cast(1 as boolean))),"c_date",cast('2020-02-02' as date)),cast('2020-02-03' as date),named_struct("c_map",map(cast('1010-01-01' as date),array(cast(1 as boolean),cast(0 as boolean),cast(null as boolean))),"c_date",cast('2020-02-02' as date)),cast('2020-02-04' as date),named_struct("c_map",map(cast('1010-01-01' as date),array(cast(null as boolean),cast(1 as boolean))),"c_date",cast('2020-02-02' as date))),
 map(cast(23 as tinyint),named_struct("c_date",cast('1010-10-01' as date),"c_array",array(map(cast(1 as boolean),cast('1010-10-01' as date))),"c_boolean",cast(1 as boolean))),
 map(cast(120.23 as decimal(5,2)),map(cast(10 as int),named_struct("c_array",array(cast('1000-01-01' as date),cast('2020-02-02' as date),cast('3030-03-03' as date)),"c_string",'Korea',"c_bigint",cast(99999 as bigint)))),
 map(cast('3030-03-03 03:03:03' as timestamp),map(cast(10 as int),array(named_struct("c_date",cast('1010-01-01' as date),"c_string",'Italy',"c_boolean",cast(0 as boolean)),named_struct("c_date",cast('1313-12-12' as date),"c_string","Turky","c_boolean",cast(0 as boolean)),named_struct("c_date",cast('4040-04-04' as date),"c_string",'Russia',"c_boolean",cast(1 as boolean)))));
```

2. CREATE CATALOG IN SR:
```
create external catalog paimon_catalog
PROPERTIES(
"type" = "paimon",
"paimon.catalog.type"="hive",
"paimon.catalog.warehouse"="hdfs://***.***.***.***:9000/user/hive/warehouse",
"hive.metastore.uris" = "thrift://***.***.***.***:9083");
```

3. SELECT FROM PAIMON TABLE IN SR:
```
mysql> select * from paimon_hdfs_orc_map_nest_lz4;
ERROR 1064 (HY000): Failed to call the nextChunkOffHeap method of off-heap table scanner. java exception details: java.io.IOException: Failed to get the next off-heap table chunk of paimon.
	at com.starrocks.paimon.reader.PaimonSplitScanner.getNext(PaimonSplitScanner.java:219)
	at com.starrocks.jni.connector.ConnectorScanner.getNextOffHeapChunk(ConnectorScanner.java:101)
Caused by: java.lang.NullPointerException
	at com.starrocks.paimon.reader.PaimonColumnValue.getBoolean(PaimonColumnValue.java:45)
	at com.starrocks.jni.conne
```

fe.warn.log:
```
2023-10-24 18:07:52,303 WARN (starrocks-mysql-nio-pool-27|633) [DefaultCoordinator.getNext():732] query failed: Failed to call the nextChunkOffHeap method of off-heap table scanner. java exception details: java.io.IOException: Failed to get the next off-heap table chunk of paimon.
        at com.starrocks.paimon.reader.PaimonSplitScanner.getNext(PaimonSplitScanner.java:219)
        at com.starrocks.jni.connector.ConnectorScanner.getNextOffHeapChunk(ConnectorScanner.java:101)
Caused by: java.lang.NullPointerException
        at com.starrocks.paimon.reader.PaimonColumnValue.getBoolean(PaimonColumnValue.java:45)
        at com.starrocks.jni.connector.OffHeapColumnVector.appendValue(OffHeapColumnVector.java:483)
        at com.starrocks.jni.connector.OffHeapColumnVector.appendArray(OffHeapColumnVector.java:408)
        at com.starrocks.jni.connector.OffHeapColumnVector.appendValue(OffHeapColumnVector.java:521)
        at com.starrocks.jni.connector.OffHeapColumnVector.appendMap(OffHeapColumnVector.java:427)
        at com.starrocks.jni.connector.OffHeapColumnVector.appendValue(OffHeapColumnVector.java:528)
        at com.starrocks.jni.connector.OffHeapColumnVector.appendStruct(OffHeapColumnVector.java:438)
        at com.starrocks.jni.connector.OffHeapColumnVector.appendValue(OffHeapColumnVector.java:534)
        at com.starrocks.jni.connector.OffHeapColumnVector.appendMap(OffHeapColumnVector.java:427)
        at com.starrocks.jni.connector.OffHeapColumnVector.appendValue(OffHeapColumnVector.java:528)
        at com.starrocks.jni.connector.OffHeapTable.appendData(OffHeapTable.java:95)
        at com.starrocks.jni.connector.ConnectorScanner.appendData(ConnectorScanner.java:86)
        at com.starrocks.paimon.reader.PaimonSplitScanner.getNext(PaimonSplitScanner.java:209)
        ... 1 more
```

### Expected behavior (Required)
return the right result

### Real behavior (Required)
error

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
```
branch-3.2-358a620
```

--- a/java-extensions/jni-connector/src/main/java/com/starrocks/jni/connector/ColumnType.java
+++ b/java-extensions/jni-connector/src/main/java/com/starrocks/jni/connector/ColumnType.java
@@ -60,10 +60,10 @@ public class ColumnType {
     List<Integer> fieldIndex;
     String rawTypeValue;
 
+    int precision = -1;
     int scale = -1;
     private static final Map<String, TypeValue> PRIMITIVE_TYPE_VALUE_MAPPING = new HashMap<>();
     private static final Map<TypeValue, Integer> PRIMITIVE_TYPE_VALUE_SIZE = new HashMap<>();
-
     private static final Map<TypeValue, String> PRIMITIVE_TYPE_VALUE_STRING_MAPPING = new HashMap<>();
 
     private static final int MAX_DECIMAL32_PRECISION = 9;
@@ -184,14 +184,15 @@ public class ColumnType {
     }
 
     private void parse(StringScanner scanner) {
-        int p;
-        if (scanner.s.startsWith("decimal")) {
-            p = scanner.s.length();
-        } else {
-            p = scanner.indexOf('<', ',', '>');
-        }
+        int p = scanner.indexOf('<', ',', '>');
+        int end = scanner.indexOf(')') + 1;
         String t = scanner.substr(p);
-        scanner.moveTo(p);
+        if (t.startsWith("decimal")) {
+            t = scanner.substr(end);
+            scanner.moveTo(end);
+        } else {
+            scanner.moveTo(p);
+        }
         // assume there is no blank char in `type`.
         typeValue = null;
         switch (t) {

--- a/java-extensions/paimon-reader/src/main/java/com/starrocks/paimon/reader/PaimonColumnValue.java
+++ b/java-extensions/paimon-reader/src/main/java/com/starrocks/paimon/reader/PaimonColumnValue.java
@@ -151,7 +151,11 @@ public class PaimonColumnValue implements ColumnValue {
 
     private void toPaimonColumnValue(List<ColumnValue> values, InternalArray array, DataType dataType) {
         for (int i = 0; i < array.size(); i++) {
-            PaimonColumnValue cv = new PaimonColumnValue(InternalRowUtils.get(array, i, dataType), dataType);
+            PaimonColumnValue cv = null;
+            Object o = InternalRowUtils.get(array, i, dataType);
+            if (o != null) {
+                cv = new PaimonColumnValue(o, dataType);
+            }
             values.add(cv);
         }
     }

--- a/java-extensions/paimon-reader/src/main/java/com/starrocks/paimon/reader/PaimonSplitScanner.java
+++ b/java-extensions/paimon-reader/src/main/java/com/starrocks/paimon/reader/PaimonSplitScanner.java
@@ -36,7 +36,6 @@ import org.apache.paimon.table.Table;
 import org.apache.paimon.table.source.ReadBuilder;
 import org.apache.paimon.table.source.Split;
 import org.apache.paimon.types.DataType;
-import org.apache.paimon.types.DecimalType;
 import org.apache.paimon.types.RowType;
 import org.apache.paimon.utils.InternalRowUtils;
 
@@ -135,9 +134,6 @@ public class PaimonSplitScanner extends ConnectorScanner {
             DataType dataType = table.rowType().getTypeAt(index);
             String type = PaimonTypeUtils.fromPaimonType(dataType);
             requiredTypes[i] = new ColumnType(type);
-            if (dataType instanceof DecimalType) {
-                requiredTypes[i].setScale(((DecimalType) dataType).getScale());
-            }
             logicalTypes[i] = dataType;
         }
 

--- a/java-extensions/paimon-reader/src/main/java/com/starrocks/paimon/reader/PaimonTypeUtils.java
+++ b/java-extensions/paimon-reader/src/main/java/com/starrocks/paimon/reader/PaimonTypeUtils.java
@@ -37,8 +37,6 @@ import java.util.stream.Collectors;
 
 /** Convert paimon type to hive string representation. */
 public class PaimonTypeUtils {
-    public static final int MAX_DECIMAL32_PRECISION = 9;
-    public static final int MAX_DECIMAL64_PRECISION = 18;
     private PaimonTypeUtils() {}
 
     public static String fromPaimonType(DataType type) {
@@ -66,15 +64,7 @@ public class PaimonTypeUtils {
         }
 
         public String visit(DecimalType decimalType) {
-            // convert decimal(x,y) to decimal
-            int precision = decimalType.getPrecision();
-            if (precision <= MAX_DECIMAL32_PRECISION) {
-                return "decimal32";
-            } else if (precision <= MAX_DECIMAL64_PRECISION) {
-                return "decimal64";
-            } else {
-                return "decimal128";
-            }
+            return String.format("decimal(%d,%d)", decimalType.getPrecision(), decimalType.getScale());
         }
 
         public String visit(TinyIntType tinyIntType) {

